hal/src/nRF52840/usart_hal.cpp

```diff
@@ -221,54 +221,81 @@ class Usart {
         nrf_uarte_enable(uarte_);
 
         config_ = conf;
-        enabled_ = true;
+        state_ = HAL_USART_STATE_ENABLED;
 
         enableTimer();
         startReceiver();
 
-        return 0;
+        return SYSTEM_ERROR_NONE;
     }
 
     int end() {
-        CHECK_TRUE(enabled_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(state_ != HAL_USART_STATE_DISABLED, SYSTEM_ERROR_INVALID_STATE);
         AtomicSection lk;
 
-        disableInterrupts();
-
-        stopTransmission();
-        stopReceiver();
-
-        nrf_uarte_disable(uarte_);
+        CHECK(disable());
 
-        nrf_uarte_txrx_pins_disconnect(uarte_);
-        nrf_uarte_hwfc_pins_disconnect(uarte_);
-
-        HAL_Pin_Mode(txPin_, INPUT);
+        // Configuring the pin mode to PIN_MODE_NONE will disconnect input buffer to enable power savings
+        HAL_Pin_Mode(txPin_, PIN_MODE_NONE);
         HAL_Set_Pin_Function(txPin_, PF_NONE);
-        HAL_Pin_Mode(rxPin_, INPUT);
+        HAL_Pin_Mode(rxPin_, PIN_MODE_NONE);
         HAL_Set_Pin_Function(rxPin_, PF_NONE);
-
-        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
-            HAL_Pin_Mode(ctsPin_, INPUT);
-            HAL_Set_Pin_Function(ctsPin_, PF_NONE);
-        }
         if (config_.config & SERIAL_FLOW_CONTROL_RTS) {
-            HAL_Pin_Mode(rtsPin_, INPUT);
+            HAL_Pin_Mode(rtsPin_, PIN_MODE_NONE);
             HAL_Set_Pin_Function(rtsPin_, PF_NONE);
         }
+        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
+            HAL_Pin_Mode(ctsPin_, PIN_MODE_NONE);
+            HAL_Set_Pin_Function(ctsPin_, PF_NONE);
+        }
 
-        nrfx_prs_release(uarte_);
+        state_ = HAL_USART_STATE_DISABLED;
+        transmitting_ = false;
+        receiving_ = 0;
+        config_ = {};
+        rxBuffer_.reset();
+        txBuffer_.reset();
 
-        disableTimer();
+        return SYSTEM_ERROR_NONE;
+    }
 
-        enabled_ = false;
-        config_ = {};
+    int suspend() {
+        CHECK_TRUE(isEnabled(), SYSTEM_ERROR_NONE);
```

Comment:
I'd prefer to return `SYSTEM_ERROR_INVALID_STATE` here, after calling suspend() API, the user would think the UART is in the suspended state, but actually it is in the disabled state which is confusing. The suspend operation should only occur when the device is in the enabled state.